### ISSUE TITLE
AL_EFFECT_EAXREVERB performance issues

### ISSUE THREAD

**Issue by LordOfDragons** [kcat/openal-soft#902]
When enabling several AL_EFFECT_EAXREVERB effects sounds starts skipping, especially positionless sounds without any filter nor effects (like music). Is there a way to tune performance? Is the AL_EFFECT_REVERB faster than AL_EFFECT_EAXREVERB or the same?

**Comment by LordOfDragons** [comment]
I did some more testing and I ended up right now with a maximum of 16 effectslots that had been dynamically created but only 4-8 or so are actively used. How can 8 concurrent eax reverb slots cause stuttering of non-effect playing source on a strong, recent CPU?

**Comment by kcat** [comment]
Effect slots are processed even if they don't have anything connected to them, since there can be residual output, and it would add to the cost of the effect to try to detect when there's no more input and output to stop processing. Reverb also isn't a particularly cheap effect. This cost may also be higher if you're constantly changing certain parameters, since changes to certain parameters may cause a pipeline transition where it will temporarily process two reverbs internally, to avoid an abrupt transition of various feedback loops and filters.

To keep things more performant, I'd avoid using more than 4 reverb effect slots at a time, representing the 4 most important environments around the listener at a given time, and only very occasionally change diffusion, density, decay, modulation, and HF/LF reference parameters (and try not to change these parameters on multiple reverbs at once, to avoid having multiple reverbs in a costly pipeline transition at the same time).

AL_EFFECT_REVERB and AL_EFFECT_EAXREVERB are the same with regards to performance. Aside from certain parameters being fixed and unchangeable with AL_EFFECT_REVERB, the actual processing is the same.

**Comment by LordOfDragons** [comment]
This is going to be tricky. I don't have classic environment zones but instead a ray-tracer tracing sound rays between listener and sound sources calculating parameters for an effect processor to fake the environment (I don't want to go down the IR convolution route as it is costly). This in particular means a couple of things:
- the count of in-flight effect slots depends on the count of sounds that are audible
- the effect parameters for each sound are updated during each audio frame update

But for testing purpose I did once only set the eax-reverb effect type and left the parameters untouched. According to your statement not changing the effect parameters should not cause any stalling due to pipeline change. Yet still the same stuttering happens. If I do not set the eax-reverb effect type the stuttering stops. But it does not happen constantly. The stalling happens sometimes.

What kind of actions all cause a stalling or potential performance problem connected with effect slots? Maybe I see something that could trigger this problem.

**Comment by kcat** [comment]
For performance, the main thing is that all the processing is done in series. For the listener, each effect slot, and each active source, it checks for and applies any property updates to calculate the new mixing/processing parameters, one after another. Then for each active source, it resamples as needed, then mixes to the main output buffer and each send that has a target. Then for each effect slot, it processes the effect. Then it does any post-process on the main output mix, applies gain limiting and distance compensation, then writes to the device buffer.

The actual mixing process doesn't take advantage of multiple cores (though multiple cores are helpful to ensure the mixing thread has one immediately available when it needs to wake up and work), so the primary factor will be single-threaded performance. This is normally fine, even a moderately old CPU should be able to handle a lot of sources with an effect or two. But certain effects are the more costly part of the pipeline, so each extra effect slot with reverb will add more cost. Stuff like HRTF also adds notable cost to sources, as does high device sample rates (and especially HRTF with high sample rates, which disproportionately adds cost).

So one extra thing to check is to make sure the device is set to something normal, like 44 or 48khz, and not 96 or 192khz. If you're on headphones, you can also try disabling OpenAL Soft's HRTF to see if that helps with the stuttering (if it does, there are alternative options to get HRTF without as much processing cost).

> But for testing purpose I did once only set the eax-reverb effect type and left the parameters untouched. According to your statement not changing the effect parameters should not cause any stalling due to pipeline change. Yet still the same stuttering happens.

Not changing the effect parameters shouldn't cause any additional CPU use from a pipeline transition. If the audio's still stuttering, that means the CPU is having trouble keeping up with just the normal processing of the effects and sources.

**Comment by LordOfDragons** [comment]
Is there a way to debug the time usage of OpenAL?

**Comment by LordOfDragons** [comment]
I've some some more experiments. I tried reducing amount of effects playing at the same time and various other things but the stuttering remained. Then I created some openal logs and noticed it uses "PipeWire" as backend. I then tried forcing it to use "Pulse" (using .alsoftrc). Interestingly using the pulse backend I did not managed to trigger the stuttering in places where with "PipeWire" I got a bunch of stuttering. This had been just a first test and I need to do more but could it be the pipewire backend is not working well? Is there a way at run-time (when creating the context or later on) to drop a backend or force a backend? Having players to fiddle with a config file is not such a nice thing I think.

**Comment by kcat** [comment]
> Is there a way to debug the time usage of OpenAL?

Under Linux, `perf` should be able to help (though to get the most out of it, it helps to have a build that keeps debug symbols, e.g. `RelWithDebInfo`).

> This had been just a first test and I need to do more but could it be the pipewire backend is not working well?

The pipewire backend does the audio processing in real-time; when the hardware interrupt triggers, pipewire wakes up and alerts playing streams that it needs more audio, and those streams then use a real-time priority thread to provide their audio that gets mixed into the hardware buffer. Rendering the audio at that point on-demand helps lower the playback latency, as its the minimal amount of time between audio being rendered and getting to the hardware, but does make it a bit more sensitive to underruns since it has to finish in time for the hardware to play it. So far, I haven't seen much of a problem with this, though if it's a common enough problem, it can be looked into changing.

The pulseaudio backend sets up a stream with a buffer queue, where OpenAL Soft provides audio to the queue at a more leisurely pace, and the server only needs to grab audio that's already stored in the buffer queue. This gives OpenAL Soft more time to do its work, at the cost of more latency since the audio is rendered well ahead of when the hardware needs it.

You can make the pipewire backend work in a non-real-time manner by setting
```
[pipewire]
rt-mix = false
```
in .alsoftrc, which will make it use a buffer queue in a similar way. Increasing latency for slightly more underrun protection.

> Is there a way at run-time (when creating the context or later on) to drop a backend or force a backend?

No, since it's an implementation detail. Different systems may have different backends available (some may not have pulseaudio, some might not have pipewire, some might not have either pulseaudio or pipewire, etc), different users may have different experiences with different backends (one person can have an issue with pipewire and not pulseaudio, another can have an issue with pulseaudio and not pipewire). Issues with a backend in one version may be fixed in a later version. New backends may be added and old ones removed in the future.

OpenAL Soft tries to provide a good default out-of-the-box experience, and the backend only needs to do the minimal amount of work to render audio and get it to the output stream. If one app has problems on a given system, other apps are likely to as well, and if multiple systems have the same problem, it's likely there's a deeper issue to address in OpenAL Soft itself. In both cases, individual apps having control over the backend doesn't make much sense.

**Comment by LordOfDragons** [comment]
I did some more tests and the Pulse backend also starts stuttering in certain places. With PipeWire the stuttering happens already at 2-4 effects whereas with Pulse it seems to requires 8-16 or so. The problem just shifted up. So it is still a problem with the effect slots stuttering even on recent strong CPUs.

**Comment by kcat** [comment]
Are you able to provide logs when using PipeWire and PulseAudio? There could be something with the way playback is being configured that has more performance impact, which the logs may help show.

**Comment by LordOfDragons** [comment]
Today the results are again a bit different. In this test case pulse tends to scratch and become silent for a short time while pipewire stutters a bit. In the other test case pipewire had been worse. I've attached the logs of the game engine (initialization) and openal for both backends. The count of released effect slots can be misleading. If not used they are disabled (NULL_EFFECT).

[Engine logs](https://pastebin.com/sdrVCiBL), [OpenAL with Pipewire](https://pastebin.com/aubRCtfG), [OpenAL with pulse](https://pastebin.com/CguUFnY8)

**Comment by kcat** [comment]
> ```
> - Maximum Mono Sources = 8192
> - Maximum Stereo Sources = 8192
> ...
> - HRTF Enabled = 1
> ```

This is a lot of sources to have with HRTF enabled. How many sources are actually playing at one time? Does the stuttering correspond to the number of sources actively playing?

**Comment by LordOfDragons** [comment]
Right now at most 10 concurrently (can be a bit more in some situations). I just used a larger number to be on the safe side.

**Comment by kcat** [comment]
That's not a high number that I'd expect to cause a problem. But it is possible to improve performance with HRTF by setting
```
[general]
hrtf-mode=ambi1 # or ambi2 or ambi3
```
That will reduce preciseness of 3D sounds (the placement of sounds won't be as sharp), but it replaces the per-source HRTF mixing with normal mixing, and does the HRTF as a fixed post-process, which is overall cheaper. `ambi2` and `ambi3` improve placement clarity but at an increasing fixed cost.

**Comment by LordOfDragons** [comment]
This certainly helps to push the boundaries a bit. Can this be set using API somehow?

**Comment by kcat** [comment]
No, it's an implementation detail, and various modes can be added and removed as things change internally, which would cause compatibility problems if it was exposed in the API. I've considered making `ambi2` or `ambi3` the default (as it would fit better into the render pipeline, and has better CPU efficiency), with users able to opt to using `full` or something else depending on their system capabilities, but I'm not sure if it would be a welcome change since it is different from how it's been since HRTF was added (and technically not as good at positional accuracy).
